pyproject.toml

```diff
@@ -112,18 +112,36 @@ exclude_lines = [
 [tool.tox]
 legacy_tox_ini = """
 [tox]
+envlist = lint,pytest
 isolated_build = True
-envlist = site
+labels =
+  venv = py311-lint, py3{6,7,8,9,10,11,12}-pytest
 
+# The default test environments use system packages and never PyPI.
+[testenv:{lint,pytest}]
+sitepackages = True
+install_command = python3 -m pip install --no-index --no-build-isolation {opts} {packages}
+wheel_build_env = pkg
+
+# All other environments (names like py311-lint, py36-pytest, etc) are isolated
+# from the system and get their packages from PyPI, according to the specific
+# test environment being requested.  We build the wheel in a common environment.
 [testenv]
-deps=
+package = wheel
+wheel_build_env = venv-pkg
+skip_install = lint: True
+deps =
+  lint: mypy
+  lint: pycodestyle
+  lint: ruff
+  lint: vulture
+  pytest
   pytest-asyncio
-  pytest-cov
-  pytest-timeout
-commands=
-  python3 -m pytest -opythonpath= --cov {posargs}
-
-[testenv:site]
-sitepackages = True
-deps=
+  pytest: pytest-cov
+  pytest: pytest-timeout
+  pytest: pytest-xdist
+allowlist_externals = test/static-code
+commands =
+  pytest: python3 -m pytest -opythonpath= {posargs}
+  lint: test/static-code --tap
```

Comment:
Wouldn't it make a bit more sense to run this in the default test env against the system packages? Not only does this sound prone to more spontaneous failure (if e.g. ruff or mypy update on pypi), but it would also needlessly run 7 times in each venv?

Reply:
By default if you type `tox` or `tox p` then you get the system environment.  If you want to run venvs you have to explicitly select them by either `tox -e py36-pytest` or `tox -m venv`.

In case you do `tox -m venv` then it runs `pytest` for each of `py36` through `py312` but only runs `lint` on `py311`.

Reply:
Okay, so that's apparently some magic that isn't clear from just reading these declarations. Thanks!